File: sdks/oracles/ts/src/pda.ts
/**
 * Program-derived-address (PDA) derivation for the Kassandra program.
 *
 * Every seed list here is the program's PUBLIC CONTRACT, verified against the
 * processors and the test harness `*_pda` helpers in
 * `programs/oracles/tests/common/mod.rs`. Seed-byte encodings:
 *   - literal seeds (`"oracle"`, `"protocol"`, ...) are ASCII byte strings;
 *   - pubkey seeds are the 32 RAW bytes of the address;
 *   - the oracle `nonce` is a **u64 little-endian, 8 bytes**;
 *   - `content_hash` is its 32 raw bytes.
 *
 * web3.js@3.0.0-rc.2 exposes only the ASYNC `Address.findProgramAddress`
 * (there is NO `findProgramAddressSync` in this version), so every derivation
 * here is async and returns `{ address, bump }`.
 */
import { Address } from "@solana/web3.js";

import { u64LE } from "./bytes.js";
import { ATA_PROGRAM_ID, KASSANDRA_PROGRAM_ID, TOKEN_PROGRAM_ID } from "./constants.js";

/** Anything that can name an account: a web3.js `Address`/`PublicKey` or a base58 string. */
export type AddressInput = Address | string;

/** A derived PDA: its address and the canonical bump seed. */
export interface Pda {
  address: Address;
  bump: number;
}

const enc = new TextEncoder();

/** 32 raw bytes of an address (the seed form of a pubkey). */
function pubkeyBytes(a: AddressInput): Uint8Array {
  return (a instanceof Address ? a : new Address(a)).toBytes();
}

async function derive(seeds: Array<Uint8Array>, programId: Address = KASSANDRA_PROGRAM_ID): Promise<Pda> {
  const [address, bump] = await Address.findProgramAddress(seeds, programId);
  return { address, bump };
}

/** Protocol singleton PDA — seeds `[b"protocol"]`. */
export function protocol(programId?: Address): Promise<Pda> {
  return derive([enc.encode("protocol")], programId);
}

/** KASS mint-authority PDA — seeds `[b"mint_authority"]`. */
export function mintAuthority(programId?: Address): Promise<Pda> {
  return derive([enc.encode("mint_authority")], programId);
}

/** Oracle PDA — seeds `[b"oracle", nonce_u64_le]`. */
export function oracle(nonce: bigint | number, programId?: Address): Promise<Pda> {
  return derive([enc.encode("oracle"), u64LE(nonce)], programId);
}

/** Oracle stake-vault PDA (KASS token account) — seeds `[b"vault", oracle]`. */
export function stakeVault(oracleAddr: AddressInput, programId?: Address): Promise<Pda> {
  return derive([enc.encode("vault"), pubkeyBytes(oracleAddr)], programId);
}

/** Oracle-metadata PDA — seeds `[b"oracle_meta", oracle]`. Holds the plaintext
 *  subject + option labels + uri/uri_hash written by `write_oracle_meta`. */
export function oracleMeta(oracleAddr: AddressInput, programId?: Address): Promise<Pda> {
  return derive([enc.encode("oracle_meta"), pubkeyBytes(oracleAddr)], programId);
}

/** Proposer PDA — seeds `[b"proposer", oracle, authority]`. */
export function proposer(
  oracleAddr: AddressInput,
  authority: AddressInput,
  programId?: Address,
): Promise<Pda> {
  return derive([enc.encode("proposer"), pubkeyBytes(oracleAddr), pubkeyBytes(authority)], programId);
}

/** Fact PDA — seeds `[b"fact", oracle, content_hash]`. `contentHash` is 32 bytes. */
export function fact(
  oracleAddr: AddressInput,
  contentHash: Uint8Array,
  programId?: Address,
): Promise<Pda> {
  return derive([enc.encode("fact"), pubkeyBytes(oracleAddr), contentHash], programId);
}

/** FactVote PDA — seeds `[b"vote", fact, voter]`. */
export function factVote(
  factAddr: AddressInput,
  voter: AddressInput,
  programId?: Address,
): Promise<Pda> {
  return derive([enc.encode("vote"), pubkeyBytes(factAddr), pubkeyBytes(voter)], programId);
}

/** AiClaim PDA — seeds `[b"claim", oracle, proposer]`. */
export function aiClaim(
  oracleAddr: AddressInput,
  proposerAddr: AddressInput,
  programId?: Address,
): Promise<Pda> {
  return derive([enc.encode("claim"), pubkeyBytes(oracleAddr), pubkeyBytes(proposerAddr)], programId);
}

/** Market PDA — seeds `[b"market", ai_claim]`. */
export function market(aiClaimAddr: AddressInput, programId?: Address): Promise<Pda> {
  return derive([enc.encode("market"), pubkeyBytes(aiClaimAddr)], programId);
}

/** Challenger USDC escrow PDA (SPL token account) — seeds `[b"challenge_usdc", market]`. */
export function challengeUsdcVault(marketAddr: AddressInput, programId?: Address): Promise<Pda> {
  return derive([enc.encode("challenge_usdc"), pubkeyBytes(marketAddr)], programId);
}

/**
 * SPL associated-token-account address — seeds `[owner, TOKEN_PROGRAM, mint]`
 * under the {@link ATA_PROGRAM_ID}. `sweep_oracle`'s DAO treasury is
 * `ATA(dao_authority, kass_mint)`, derived exactly as the program does in
 * `processor/sweep_oracle.rs`. NOTE: derived under the ATA program, NOT the
 * Kassandra program — there is no `programId` override.
 */
export function associatedTokenAccount(
  owner: AddressInput,
  mint: AddressInput,
): Promise<Pda> {
  return derive([pubkeyBytes(owner), TOKEN_PROGRAM_ID.toBytes(), pubkeyBytes(mint)], ATA_PROGRAM_ID);
}

File: sdks/oracles/rust/src/pda.rs
//! Program-derived address helpers.
//!
//! These seed conventions are part of the program's public contract; downstream
//! code MUST derive with exactly these seeds. Each function returns
//! `(address, bump)` like [`solana_pubkey::Pubkey::find_program_address`].

use kassandra_oracles_program::config::MINT_AUTHORITY_SEED;
use solana_pubkey::Pubkey;

use crate::{ATA_PROGRAM_ID, TOKEN_PROGRAM_ID};

/// Oracle PDA — seeds `[b"oracle", nonce_le]`.
pub fn oracle(program_id: &Pubkey, nonce: u64) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"oracle", &nonce.to_le_bytes()], program_id)
}

/// Oracle-metadata PDA — seeds `[b"oracle_meta", oracle]`. Holds the plaintext
/// subject + option labels + uri/uri_hash written by `write_oracle_meta`.
pub fn oracle_meta(program_id: &Pubkey, oracle: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"oracle_meta", oracle.as_ref()], program_id)
}

/// Protocol singleton PDA — seeds `[b"protocol"]`.
pub fn protocol(program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"protocol"], program_id)
}

/// KASS mint-authority PDA — seeds `[MINT_AUTHORITY_SEED]`. Handed to the KASS
/// mint so the program's emission `MintTo` can program-sign.
pub fn mint_authority(program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[MINT_AUTHORITY_SEED], program_id)
}

/// Stake-vault PDA for an oracle — seeds `[b"vault", oracle]`. An SPL token
/// account on the KASS mint whose authority is the oracle PDA.
pub fn stake_vault(program_id: &Pubkey, oracle: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"vault", oracle.as_ref()], program_id)
}

/// Challenger USDC escrow-vault PDA for a market — seeds `[b"challenge_usdc", market]`.
pub fn challenge_usdc_vault(program_id: &Pubkey, market: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"challenge_usdc", market.as_ref()], program_id)
}

/// Proposer PDA — seeds `[b"proposer", oracle, authority]`.
pub fn proposer(program_id: &Pubkey, oracle: &Pubkey, authority: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"proposer", oracle.as_ref(), authority.as_ref()],
        program_id,
    )
}

/// Fact PDA — seeds `[b"fact", oracle, content_hash]`.
pub fn fact(program_id: &Pubkey, oracle: &Pubkey, content_hash: &[u8; 32]) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"fact", oracle.as_ref(), content_hash.as_ref()],
        program_id,
    )
}

/// FactVote PDA — seeds `[b"vote", fact, voter]`.
pub fn vote(program_id: &Pubkey, fact: &Pubkey, voter: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"vote", fact.as_ref(), voter.as_ref()], program_id)
}

/// AiClaim PDA — seeds `[b"claim", oracle, proposer]`.
pub fn ai_claim(program_id: &Pubkey, oracle: &Pubkey, proposer: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[b"claim", oracle.as_ref(), proposer.as_ref()], program_id)
}

/// The canonical KASS associated-token-account of `owner` — where the DAO
/// treasury lives. Derived under the ATA program from `[owner, token_program, mint]`.
pub fn kass_ata(owner: &Pubkey, kass_mint: &Pubkey) -> Pubkey {
    Pubkey::find_program_address(
        &[
            owner.as_ref(),
            TOKEN_PROGRAM_ID.as_ref(),
            kass_mint.as_ref(),
        ],
        &ATA_PROGRAM_ID,
    )
    .0
}
